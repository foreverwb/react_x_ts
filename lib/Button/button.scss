@import "../_main.scss";
.ui-btn {
 display: inline-flex;
 font-weight: 400;
 white-space: nowrap;
 text-align: center;
 padding: 0 16px;
 font-size: inherit;
 border-radius: 4px;
 cursor: pointer;
 box-sizing: border-box;
 vertical-align: middle;
 justify-content: center;
 align-items: center;
 border: none;
 &:disabled {
  cursor: not-allowed;
 }
 &-default {
  border: 1px solid $border-color;
  color: $text-color;
  background: #ffffff;
  &:focus, &:hover {
   color: $main-color;
   border-color: $main-color;
  }
  &:disabled {
   border-color: lighten($border-color, 20%);
   color: lighten($text-color, 20%);
   background-color: lighten($border-color, 38%);
  }
 }
 &-primary {
  border-color: $main-color;
  background-color: $main-color;
  color: #ffffff;
  &:focus, &:hover {
   background-color: darken($main-color, 6%);
  }
  &:disabled {
   border-color: lighten($main-color, 15%);
   background-color: lighten($main-color, 15%);
  }
 }
 &-danger {
  color: #f5222d;
  background-color: #f5f5f5;
  border-color: darken($border-color, 20%);
   &:focus, &:hover {
    color: #fff;
    background-color: #ff4d4f;
    border-color: #ff4d4f
   }
   &:disabled {
    color: #f5222d;
    background-color: darken(#f5f5f5, 6%);
   }
 }

}
.ui-btn-loadingIcon {
 &-animationStart {
  margin-left: -16px !important;
  opacity: 0;
 }
 &-animationEnd {
  margin-left: 0 !important;
  opacity: 1;
  transition: margin-left 200ms, opacity 200ms;
 }
}
.ui-btn + .ui-btn {
 margin-left: 6px;
}