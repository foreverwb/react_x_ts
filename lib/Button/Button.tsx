import React, {Fragment} from 'react';
import {scopedClassMaker} from '../untils/classes';
import Icon from '../Icon/Icon';
import './button.scss';

const sc = scopedClassMaker('ui-btn');
const sc_icon = scopedClassMaker('ui-icon');

interface IProps {
  level: string;
  type?: 'button' | 'submit' | 'reset';
  disabled?: boolean;
  icon?: string;
  iconPosition?: 'left' | 'right';
  loading?: boolean;
  onClick?: React.MouseEventHandler<HTMLButtonElement>
}

const Button: React.FunctionComponent<IProps> = (props) => {
  const icon = props.icon && <Icon name={props.icon}/>;
  const loadingIcon = (
    <Icon name="loading" className={sc_icon('loading')}/>
  );
  const iconWrapper = props.loading ? loadingIcon : icon;
  const content = typeof props.children === 'string' ? <span>{props.children}</span> : props.children;
  const inner = (
    props.iconPosition === 'left' ?
      <Fragment>{iconWrapper}{content}</Fragment>
      :
      <Fragment>{content}{iconWrapper}</Fragment>
  );

  const onClick = (e: React.MouseEvent) => {
    if (props.disabled) return e.preventDefault();
    props.onClick && props.onClick.call(null, e.target, e);
  };

  return (
    <button
      className={sc({'': true, [props.level]: true})}
      type={props.type}
      disabled={props.disabled}
      onClick={onClick}>
      {inner}
    </button>
  );
};
Button.defaultProps = {
  disabled: false,
  type: 'button',
  iconPosition: 'left'
};
export default Button;