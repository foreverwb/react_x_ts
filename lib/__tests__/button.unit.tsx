import React from 'react';
import Button from '../Button';
import renderer from 'react-test-renderer'

describe('button', () => {
  it('is btn', () => {
    const json = renderer.create(<Button/>).toJSON()
    expect(json).toMatchSnapshot()
  })
})