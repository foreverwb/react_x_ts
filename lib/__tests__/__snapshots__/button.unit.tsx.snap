// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button is btn 1`] = `
<div>
  Button
</div>
`;
